File: C_src/Easy/504_Base_7/Base_7_504.c
#include <stdio.h>
#include <stdlib.h>

// Use recursive:
// Time Complexity: O(logn)
// Space Complexity:O(1)
// Runtime: 0 ms, faster than 100.00%
// Memory Usage: 6.7 MB, less than 100.00%
char *convertToBase7(int num)
{
    char *result = calloc(11, sizeof(char));
    if (num == 0)
    {
        return "0";
    }

    if (num < 0)
    {
        sprintf(result, "-%s", convertToBase7(abs(num)));
    }
    else
    {
        char *tmp = convertToBase7(num / 7);
        if (tmp == "0")
        {
            tmp = "";
        }
        sprintf(result, "%s%d", tmp, num % 7);
    }
    return result;
}

File: Golang_src/easy/base7_504.go
package easy

import (
	"strconv"
)

// Use iterative:
// Time Complexity: O(logn)
// Space Complexity:O(1)
// Runtime: 0 ms, faster than 100.00%
// Memory Usage: 2 MB, less than 100.00%
func convertToBase7(num int) string {
	if num == 0 {
		return "0"
	}

	result := ""
	isNegative := false

	if num < 0 {
		num = -num
		isNegative = true
	}

	for num != 0 {
		result = strconv.Itoa(num%7) + result
		num /= 7
	}

	if isNegative {
		result = "-" + result
	}

	return result
}

// Use recursive:
// Time Complexity: O(logn)
// Space Complexity:O(1)
// Runtime: 0 ms, faster than 100.00%
// Memory Usage: 2 MB, less than 100.00%
func convertToBase72(num int) string {
	if num < 0 {
		return "-" + convertToBase72(-num)
	}
	if num < 7 {
		return strconv.Itoa(num)
	}

	return convertToBase72(num/7) + strconv.Itoa(num%7)
}
